# Act 3 Regresión lineal en finanzas

**Nombre:** [Completar]  
**Matrícula:** [Completar]  
**Fecha:** 7 de septiembre de 2026

Se estima el rendimiento mensual de una acción a partir del rendimiento mensual del mercado. Los 60 datos son simulados y reproducibles.


In [1]:
%pip install -q pandas numpy statsmodels

# Conclusión: estas son las únicas bibliotecas necesarias para ejecutar el ejercicio.

Note: you may need to restart the kernel to use updated packages.


In [2]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

rng = np.random.default_rng(20260907)
n = 60
dataset = pd.DataFrame({
    "Fecha": pd.date_range("2021-01-31", periods=n, freq="ME"),
    "Rendimiento_mercado": rng.normal(0.007, 0.04, n)
})
dataset["Rendimiento_accion"] = (
    0.004 + 1.15 * dataset["Rendimiento_mercado"] + rng.normal(0, 0.018, n)
)
print(dataset.head().to_string(index=False))

# Conclusión: se usan 60 meses simulados; no hay datos personales ni información de una empresa real.

     Fecha  Rendimiento_mercado  Rendimiento_accion
2021-01-31             0.031650            0.012530
2021-02-28            -0.023146            0.014286
2021-03-31            -0.092070           -0.082731
2021-04-30            -0.030007           -0.032021
2021-05-31             0.010324            0.003063


In [3]:
X = sm.add_constant(dataset[["Rendimiento_mercado"]])
Y = dataset["Rendimiento_accion"]
modelo = sm.OLS(Y, X).fit()
print(modelo.summary())

# Conclusión: el modelo calcula intercepto, coeficiente, R2, R2 ajustado y p-values.

                            OLS Regression Results                            
Dep. Variable:     Rendimiento_accion   R-squared:                       0.857
Model:                            OLS   Adj. R-squared:                  0.854
Method:                 Least Squares   F-statistic:                     346.7
Date:                Mon, 07 Sep 2026   Prob (F-statistic):           3.83e-26
Time:                        16:23:08   Log-Likelihood:                 155.79
No. Observations:                  60   AIC:                            -307.6
Df Residuals:                      58   BIC:                            -303.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   0.0054    

## Resultados e interpretación

- Intercepto: **0.005367**.
- Coeficiente del mercado: **1.114873**. Cuando el mercado aumenta 1 punto porcentual, el rendimiento esperado de la acción cambia aproximadamente 1.115 puntos porcentuales.
- R2: **0.8567**. El modelo explica cerca de 85.7% de la variación observada en esta muestra.
- El p-value del mercado es **3.83e-26**, menor que 0.05, por lo que la variable es estadísticamente significativa al 5%.
- La asociación no demuestra causalidad y los datos simulados no sirven para recomendar inversiones.


In [4]:
excel = {
    "Intercepto": np.float64(0.005366966703811423),
    "Coeficiente mercado": np.float64(1.1148727290969567),
    "R2": 0.8566824676212552,
    "R2 ajustado": 0.8542114756836907,
    "p-value mercado": np.float64(3.8319664254351636e-26)
}
python = {
    "Intercepto": modelo.params["const"],
    "Coeficiente mercado": modelo.params["Rendimiento_mercado"],
    "R2": modelo.rsquared,
    "R2 ajustado": modelo.rsquared_adj,
    "p-value mercado": modelo.pvalues["Rendimiento_mercado"]
}
comparacion = pd.DataFrame({"Excel": excel, "Python": python})
comparacion["Diferencia"] = (comparacion["Excel"] - comparacion["Python"]).abs()
print(comparacion.to_string())

# Conclusión: Excel y Python producen los mismos resultados; las diferencias posibles son solo de redondeo.

                            Excel        Python    Diferencia
Intercepto           5.366967e-03  5.366967e-03  8.673617e-19
Coeficiente mercado  1.114873e+00  1.114873e+00  0.000000e+00
R2                   8.566825e-01  8.566825e-01  1.110223e-16
R2 ajustado          8.542115e-01  8.542115e-01  0.000000e+00
p-value mercado      3.831966e-26  3.831966e-26  1.056108e-39


## Excel y Python

| Criterio | Excel | Python |
|---|---|---|
| Facilidad | Visual y rápido | Requiere código |
| Automatización | Limitada sin macros | Alta |
| Reproducibilidad | Depende de rangos y opciones | Alta con código y semilla |
| Datos grandes | Menos conveniente | Más escalable |
| Mejor uso | Análisis pequeño y presentación rápida | Procesos repetitivos y modelos avanzados |

## Reflexión personal

Escribe aquí de 6 a 8 líneas sobre qué herramienta te resultó más clara, qué aprendiste al comparar los resultados y cuál usarías en un análisis financiero mensual.

## Referencias

statsmodels. (s. f.). *Ordinary Least Squares*. https://www.statsmodels.org/stable/generated/statsmodels.regression.linear_model.OLS.html
